[Infrastructure] Improve clean.{cmd|sh} by adding git clean -xdf cleanup step, unifying implementation with corefx and adding VBCScompiler.exe kill logic and

4creators opened the pull request:
`clean.{cmd|sh}` was updated based on corefx implementation. Supported commands can be run by invoking `clean.{cmd|sh}` directly or via `run clean` command on all supported platforms. New clean arguments are available:

- `-c`  allows to delete user-local NuGet packages 
- `-all` allows to run `git clean -xdf` command (available only for direct invocation of clean)

Support for `-t` switch allowing to delete `tools` directory was dropped. Additionally to Windows script VBCSCompiler.exe kill logic was added. It may be necessary to add it to sh script in the feature once build tools will start using VBCSCompiler.exe.

`config.json` defines new `global` command line settings:

- `-CleanAllProjects` which is equivalent to `-b` argument
- `-CleanPackages` which is equivalent to `-p` argument
- `-CleanPackagesCache` which is equivalnet to `-c` argument

4creators commented:
PTAL
@weshaggard @janvorli

weshaggard commented:
@4creators I think you should just simply the entire script to match the clean.cmd in corefx https://github.com/dotnet/corefx/blob/master/clean.cmd. Which already takes the VBCSCompiler into account and does git clean for the all case.

4creators commented:
@weshaggard 

I have checked that and there are several fundamental difference between scripts:

1. Execution of corefx script for most options requires installation of build tools.
2. Due to above it is impossible to clean tools directory without using `nuclear` command `git clean -xdf `  what not always is acceptable during development.

Due to above my proposal would be to use console based script with basic options (delete bin, tools, repo packages) and perhaps add only build-tools based argument _-c - Deletes the user-local nuget package cache_.

weshaggard commented:
What cases do you want to just delete the Tools directory but not everything else? I've generally found if I want to blow that away then I want the everything cleaned up. Or in the rare cases I want just the Tools directory deleted I will manually delete that directory.

4creators commented:
In my scenario I am working with custom modified build tools which I can update while building separate project, upload to `myget` and after updating `BuildToolsVersion.txt` I would like to restore only build tools while keeping all other directories intact. Or I would like to delete and update only `packages` and `Tools` directory with `bin` directory intact.

But the major problem with using `build-tools` based `clean` command like in `corefx` case is caused by the fact that it is necessary to install .NET Core and build-tools even before running simplest clean command. AFAIR this was discussed in several issues already.

weshaggard commented:
>In my scenario I am working with custom modified build tools which I can update while building separate project, upload to myget and after updating BuildToolsVersion.txt I would like to restore only build tools while keeping all other directories intact. Or I would like to delete and update only packages and Tools directory with bin directory intact.

If you update the BuildToolsVersion.txt the first think that happens is a clean of the Tools directory (https://github.com/dotnet/coreclr/blob/master/init-tools.cmd#L28) the next time you run a build command and then it restores the new Tools. You can also directly call init-tools.cmd force to get a clean tools directory. So I think that scenario is pretty well covered without baking more logic about it in clean.

>But the major problem with using build-tools based clean command like in corefx case is caused by the fact that it is necessary to install .NET Core and build-tools even before running simplest clean command. 

If you are running a clean command then you generally already have build-tools initialized, and if you don't then you are just getting yourself ready for the next command you run which will not need to initialize the tools. 

The main reason we moved the basic clean logic into msbuild is to avoid duplication of information between the scripts and the build files. The build file already knows where it puts stuff and so it knows where to clean things too.

>AFAIR this was discussed in several issues already.

Would you mind pointing me at such issues so I can get a better idea of the issues folks are having.

4creators commented:
@weshaggard 
These are two issues I have found right now but think there were other discussions as well:
https://github.com/dotnet/corefx/issues/15184
https://github.com/dotnet/corefx/issues/20665

weshaggard commented:
Those are issues talking about the run tool which is a little different then what we are discussing here. For this discussion it is about whether or not we need to initialize tools to do a basic clean which I think we need to regardless of whether or not we use the run tool, because I want msbuild to be the one to handle those basic cleans so we don't need to duplicate logic about which paths should be cleaned up.

4creators commented:
>  I think that scenario is pretty well covered without baking more logic about it in clean

Agree - didn't know that feature

> The main reason we moved the basic clean logic into msbuild is to avoid duplication of information between the scripts and the build files

Agree as well - the only problem is that build-tools have to installed.

4creators commented:
@dotnet-bot OSX10.12 x64 Checked Innerloop Build and Test

4creators commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

4creators commented:
@weshaggard 

But `clean` in `corefx` uses run tool:

https://github.com/dotnet/corefx/blob/63f536f9f70fcab7bb5bf0fdd0088558481fcd8d/clean.cmd#L19

weshaggard commented:
>But clean in corefx uses run tool:

Yes it does but it is only using it to get a call out to msbuild. It is also our current way to parse additional arguments but the main purpose is to call a msbuild clean target.

4creators commented:
@weshaggard 
OK. My understanding is that you would rather convert current clean.cmd script to equivalent of the clean in corefx. 

Would yo prefer me to close this PR and do it in separate PR or it would be enough to work in this PR and later squash commits?

weshaggard commented:
Yes my preference would be to make the clean scripts consistent and I prefer the approach in corefx. If there are scenarios that are missing I'm definitely open to addressing them.

Feel free to reuse this PR, you can squash the changes away if you like as well.

>What about legacy command which cleans only bin.

I'm not sure I understand the value with doing that? Are you just trying to avoid initializing the tools? I would expect them to already be initialized in most cases you want to just delete bin.

4creators commented:
@weshaggard 
> Are you just trying to avoid initializing the tools
No. The problem is that CMake generates a lot of files which go to `/bin/obj` and `/bin/tests` directories and they cannot be removed with msbuild. Obviously it is possible to remove 'bin' with run.cmd as well but it could be overkill and perhaps it is more efficient to do it as `special switch` -b like -all works in corefx.

weshaggard commented:
https://github.com/dotnet/coreclr/blob/master/build.proj#L22 Yes it looks like someone scoped the clean to just particular files. So we have too choices:

1. Make the msbuild clean command delete the entire bin folder like in corefx.
2. Bake to knowledge into clean script about the bin directory.

I would lean towards (1) but I could live with either of those. @jkotas and @janvorli do more work in the coreclr repo then I do so I will let them decide. Whatever we do I would like to do it consistently in both clean.cmd and clean.sh as well as doing it in corefx repo.

jkotas commented:
I only ever use `clean all` type of cleanup that kills VBCSCompiler and does `git clean -xdf`. The logic is directly in the clean script - both in CoreFX and CoreRT repos:
- https://github.com/dotnet/corefx/blob/master/clean.cmd#L5
- https://github.com/dotnet/corert/blob/master/buildscripts/clean.cmd

I think it is impossible to tell when it is a good idea to use the fine grained variants. I use clean when I get weird build break. It is always way easier to just clean everything than waste time on trying to debug the weird build break.

4creators commented on clean.cmd:
```diff
@@ -1,105 +1,39 @@
+tasklist /fi "imagename eq VBCSCompiler.exe" |find ":" > nul
+:: Compiler is running if errorlevel == 1
+if errorlevel 1 (
+	echo Stop VBCSCompiler.exe execution.
```
Executing taskkill only if at least one VBCSCompiler.exe process is running

weshaggard replied:
Does it hurt to blindly do it?

weshaggard replied:
I don't mind this logic being here if it is useful, but if it stays I would appreciate the same change going into corefx.

4creators replied:
This more effective than running through loop every time. I will introduce change to corefx as well.

4creators commented:
@dotnet-bot  test Ubuntu arm64 Cross Debug Innerloop Build

4creators commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

4creators commented:
Addressed PR feedback, CI is green, both `clean.cmd` and `clean.sh` have been updated based on corefx implementation, skipped implementation of -t switch (delete tools dir) due to lack of scenario in which this can be used and overlap with clean -all command.
PTAL
@weshaggard @jkotas @janvorli

weshaggard commented on clean.cmd:
```diff
@@ -1,105 +1,42 @@
 
-echo Running clean.cmd
+set NO_DASHES_ARG=%1
+if /I [%NO_DASHES_ARG:-=%] == [?] goto Usage
```
Thanks for adding this. Can you do the same in corefx?

4creators replied:
Yes

4creators replied:
Just checked that it is there - so it seems that rather I copied it here

weshaggard replied:
Hmmm. I don't see it in corefx https://github.com/dotnet/corefx/blob/master/clean.cmd.

4creators replied:
I deleted my comment - misread lines of code 😄
